virt-v2v/cold/customize-image.go

```diff
@@ -0,0 +1,145 @@
+	if err != nil {
+		return err
+	}
+	windowsScriptsPath := filepath.Join(DIR, "scripts", "windows")
+	initPath := filepath.Join(windowsScriptsPath, "init.bat")
+	restoreScriptPath := filepath.Join(windowsScriptsPath, "restore_config.ps1")
+
+	uploadPath := fmt.Sprintf("%s:%s", restoreScriptPath, WIN_FIRSTBOOT_SCRIPTS_PATH)
+
+	var extraArgs []string
+	extraArgs = append(extraArgs, getScriptArgs("firstboot", initPath)...)
+	extraArgs = append(extraArgs, getScriptArgs("upload", uploadPath)...)
+
+	err = CustomizeImage(diskPath, extraArgs...)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// getScriptArgs generates a list of arguments.
+//
+// Arguments:
+//   - argName (string): Argument name which should be used for all the values
+//   - values (...string): The list of values which should be joined with argument names.
+//
+// Returns:
+//   - []string: List of arguments
+//
+// Example:
+//   - getScriptArgs("firstboot", boot1, boot2) => ["--firstboot", boot1, "--firstboot", boot2]
+func getScriptArgs(argName string, values ...string) []string {
+	var args []string
+	for _, val := range values {
+		args = append(args, fmt.Sprintf("--%s", argName), val)
+	}
+	return args
+}
+
+// CustomizeImage executes `virt-customize` to customize the image.
+//
+// Arguments:
+//   - diskPath (string): The path to the disk image that is being customized.
+//   - extraArgs (...string): The additional arguments which will be appended to the `virt-customize` arguments.
+//
+// Returns:
+//   - error: An error if something goes wrong during the process, or nil if successful.
+func CustomizeImage(diskPath string, extraArgs ...string) error {
+	args := []string{"--verbose", "-a", diskPath}
+	args = append(args, extraArgs...)
+	customizeCmd := exec.Command("virt-customize", args...)
+
+	fmt.Println("exec:", customizeCmd)
+	if err := customizeCmd.Run(); err != nil {
+		return fmt.Errorf("error executing virt-customize command: %w", err)
+	}
+	return nil
+}
+
+// FindRootDiskImage locates the root disk image in the specified directory based on the disk name pattern.
+//
+// This function looks for disk files in the `dir` directory that match the pattern `*-sd*`. It uses the `V2V_RootDisk`
+// environment variable to determine the disk name, falling back to the default disk name "sda" if the environment variable is not set or invalid.
+//
+// Arguments:
+//   - dir (string): The directory containing the disk files.
+//
+// Returns:
+//   - string: The full path to the root disk image.
+//   - error: An error if the root disk image is not found or if there is an issue reading the disk directory.
+func FindRootDiskImage(dir string) (string, error) {
+	var err error
+
+	// Get all disk files matching the pattern
+	disks, err := filepath.Glob(filepath.Join(dir, "*-sd*"))
+	if err != nil {
+		return "", fmt.Errorf("error getting disks: %w", err)
+	}
+
+	// Default disk name
+	diskName := "sda"
+
+	// Check if V2V_RootDisk environment variable is set
```

Comment:
that was added to cope with a scenario of having multi-boot disks but what if there's a single disk that is bootable and it's not sda? it would be better to get the information about the bootable disk from the output of virt-v2v

Reply:
Which virt-v2v output do you mean?

Reply:
It can be scanned from the virt-v2v logs, but it will cause a different issue since the output available only after starting the conversion command, so we will have to run it twice. We had a discussion about this in the multi boot [ticket](https://issues.redhat.com/browse/MTV-1079) and Richard suggested to use `virt-v2v-inspector` prior to executing the command, its probably will be the best solution.

https://libguestfs.org/virt-v2v-inspector.1.html

Reply:
I meant the configuration of the converted VM that virt-v2v generates

Reply:
@bkhizgiy worked on:
https://issues.redhat.com/browse/MTV-1079

I asked her and she sent me this link:
https://issues.redhat.com/browse/MTV-1079?focusedId=25052893&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-25052893

we use:
`info: virt-v2v: virt-v2v 2.5.6rhel=9,release=3.el9 (x86_64)`
and this fix it on 2.5.7 :-( 

in our version we have debug output with the information:
`libguestfs: trace: v2v: inspect_get_mountpoints = ["/boot", "/dev/sda1", "/", "/dev/vg_rhev-orange-03/lv_root"]`
( IMHO it's not a good idea to pars the debug output ... _

@rwmjones also suggested using  virt-v2v-inspector.
https://issues.redhat.com/browse/MTV-1079?focusedId=25052873&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-25052873

IMHO opinion we should wait for the fix in 2.5.7, @mnecas WDYT ?

Reply:
@ahadas It can be taken from the virt-v2v final output, but again, that means we need to run a full conversion before being able to get this info for the actual run....

Reply:
> It can be taken from the virt-v2v final output,

@bkhizgiy , that will be great for this use case, because we run it after virt-v2v finish, but on my build we don't have this output ... what am i missing ?

Reply:
maybe i have old virt-v2v, v2.5.6

Reply:
@yaacov  Why can't we bump to 2.5.7 now? Are we missing the build on the u/s? I know we are using 2.6.4 in d/s so we should synchronise them.

@bkhizgiy I would consider the scraping and trying to find some data in logs our last option and would rather go other direction.

What is the difference between using the virt-v2v-inspector and checking the instance XML file? Does it show more details? I'm not that familiar with it but will definitely check it tmr.

I can't access my computer today so can't do much investigations. If possible I would try to find some solutions for which we don't need to way even if it is bit hack, we can refactor later after we will get the fix to unblock users.

Reply:
> @ahadas It can be taken from the virt-v2v final output, but again, that means we need to run a full conversion before being able to get this info for the actual run....

Note this is what virt-v2v-inspector does (a full conversion). It's not as expensive as you think because it doesn't do any copying, although it also isn't completely free either (especially for Linux guests because dracut is slow).

Reply:
As there's a bit of confusion about virt-v2v version numbers, it's a good idea to refer to these upstream branches:

master branch: https://github.com/libguestfs/virt-v2v/commits/master/

- This is the active development branch

rhel-9.5 branch: https://github.com/libguestfs/virt-v2v/commits/rhel-9.5

- Features backported to RHEL 9.5 + RHEL-only patches.  You'll have to scroll down a bit through the RHEL patches to see what we've backported.
- Since we backport aggressively, you'll find almost all upstream patches are already present in RHEL version 2.5.6.

rhel-9.4 (z-stream) branch: https://github.com/libguestfs/virt-v2v/commits/rhel-9.4

- For completeness, what's been backported to 9.4.z

Reply:
the PR we are looking for is in  9.5
https://github.com/libguestfs/virt-v2v/commit/4df3dcf64da91645d20aa9f74cd04fa8985c4610

Reply:
how do we use the rhel-9.5 rpms ?

Reply:
@mnecas hi, so if we go with `virt-v2v-inspector` 
a. we have it already installed
b. running it is "fast" compering to running `virt-v2v`

running it will look like:
`virt-v2v-inspector -i libvirtxml  /var/tmp/v2v/<vm name>.xml -o /var/tmp/v2v/<vm name>.inspect.xml`

and the output will include the root disk:
```xml
<?xml version='1.0' encoding='utf-8'?>
<v2v-inspection>
  <!-- generated by virt-v2v-inspector 2.5.6fedora=40,release=1.fc40 -->
  <program>virt-v2v-inspector</program>
  <package>virt-v2v</package>
  <version>2.5.6</version>
  <disks>
    <disk index='0'>
      <virtual-size>12884901888</virtual-size>
      <allocated estimated='true'>1530933248</allocated>
    </disk>
  </disks>
  <operatingsystem>
    <name>linux</name>
    <distro>rhel</distro>
    <osinfo>rhel8.2</osinfo>
    <arch>x86_64</arch>
    <major_version>8</major_version>
    <minor_version>2</minor_version>
    <package_format>rpm</package_format>
    <package_management>dnf</package_management>
    <product_name>Red Hat Enterprise Linux 8.2 (Ootpa)</product_name>
    <product_variant>unknown</product_variant>
    <root>/dev/vg_rhev-orange-03/lv_root</root>
    <mountpoints>
      <mountpoint dev='/dev/vg_rhev-orange-03/lv_root'>/</mountpoint>
      <mountpoint dev='/dev/sda1'>/boot</mountpoint>
    </mountpoints>
  </operatingsystem>
</v2v-inspection>
```

this is strange because in this vm we have only 1 disk `/sda`

@rwmjones what am i missing ? why i don't see `/` mounted using sda ?

Reply:
> the PR we are looking for is in 9.5 [libguestfs/virt-v2v@4df3dcf](https://github.com/libguestfs/virt-v2v/commit/4df3dcf64da91645d20aa9f74cd04fa8985c4610)

That commit is present in RHEL 9.5, but not in 9.4-z.

> how do we use the rhel-9.5 rpms ?

You'll have to ask Arik about this, but I believe there was some problem with consuming RHEL 9.5 RPMs in MTV right now.

virt-v2v is shipped using the ordinary RHEL process and schedule.  RHEL 9.5 will ship in November [technically RH-only information, but an open secret].

If you cannot use the 9.5 RPMs early then you'll need to request this to be backported to 9.4-z, which starts with a Jira ticket again RHEL product, virt-v2v component.  Nothing happens without a Jira ticket.

Since 9.4-z is a heavyweight operation involving coordinating many people I'd suggest batching requests for backports together so we don't need to go through it more than once (more).

Reply:
@rwmjones i am trying to use `virt-v2v-inspector`

i get
```xml
 <mountpoints> <mountpoint dev='/dev/vg_rhev-orange-03/lv_root'>/</mountpoint>
 <mountpoint dev='/dev/sda1'>/boot</mountpoint> </mountpoints>
```
i expected one disk `sda` ounted at `/`

what am i missing ?

Reply:
> @mnecas hi, so if we go with `virt-v2v-inspector` a. we have it already installed b. running it is "fast" compering to running `virt-v2v`
> 
> running it will look like: `virt-v2v-inspector -i libvirtxml /var/tmp/v2v/<vm name>.xml -o /var/tmp/v2v/<vm name>.inspect.xml`
> 
> and the output will include the root disk:
> 
> ```
>  <disks>
>    <disk index='0'>
>      <virtual-size>12884901888</virtual-size>
>      <allocated estimated='true'>1530933248</allocated>
>    </disk>
>  </disks>
>     <mountpoints>
>       <mountpoint dev='/dev/vg_rhev-orange-03/lv_root'>/</mountpoint>
>       <mountpoint dev='/dev/sda1'>/boot</mountpoint>
>     </mountpoints>
>   </operatingsystem>
> </v2v-inspection>
> ```
> 
> this is strange because in this vm we have only 1 disk `/sda`
> 
> @rwmjones what am i missing ? why i don't see `/` mounted using sda ?

This is correct isn't it?  It's showing there is one disk with virtual size 12GB, with about 1.5GB of data on it. With root `/` mounted on a logical volume, and `/boot` on the first partition of the disk.

`/dev/sda1` etc are actually libguestfs standard device names (https://libguestfs.org/guestfs.3.html#block-device-naming).

Inside the guest this might appear as `C:` or `/dev/vda1`.  (We actually have Linux guest device and Windows drive mapping information, but I can't remember if virt-v2v-inspector shows it).

Reply:
> Inside the guest this might appear as C: or /dev/vda1

we need the root disk to know on what disk to run `virt-customize` 
we don't care in this step what this disk is called on the guest, only how it's currently called in our pod, e.g. the file name that `virt-v2c` created in the `-os` dir

our question is how we know what `.img` to use from the files we get in the `-os` after `virt-v2v` finishes, i am missing something here, how do I get what disk include the `/etc` dir, so I can run `virt-customize` on ?

Reply:
The disks are listed in order they were specified in the source XML.

However I think it'd be possible to put the source filename into the XML too if that helps.  (Virt-v2v's internal metadata model is here: https://github.com/libguestfs/virt-v2v/blob/master/lib/types.mli).  Starts with a Jira ticket (or a pull request if you fancy implementing it).

Reply:
> how do I get what disk include the /etc dir, so I can run virt-customize on ?

I think actually there's also a misunderstanding here.  For virt-customize you should specify **all** disks of the guest (in order) using multiple `-a` options.  This is because you could have a guest which has, for example, a separate `/usr` from a second disk.  Virt-customize will sort this out if you give it all the disks, you don't have to worry about it.

Reply:
> For virt-customize you should specify all disks of the guest (in order) using multiple -a options.

WOW: 
ok, so we don't actually need to run `virt-v2v-inspector` , just look on all the `.img` files and add them with `-a`

@rwmjones last question ( i hope ) do i need to tell `virt-customize` somehow how to mount this disks, or it knows by inspecting the files ?

```basg
# is this enough ? 
# or i need something more ?  ( is the order important ?)
virt-customize -a disk-sda.img -a disk-sdb.img --run ...
```

Reply:
virt-customize uses a thing called libguestfs inspection to examine the disks and determine how they would be arranged in the original guest (using clues from `/etc/fstab`, the Windows registry, and a big pile of heuristics), it then mounts up the filesystems automatically.  More here: https://libguestfs.org/guestfs-faq.1.html#why-does-virt-cat-only-work-on-a-real-vm-image-but-virt-df-works-on-any-disk-image

However you might still want to use virt-v2v-inspector (for other reasons, maybe not here).  If you want to know in advance how much space to allocate for the target disks after virt-v2v, or you need to know other information that virt-v2v normally only provides after conversion (like whether virtio-win drivers can be installed), then virt-v2v-inspector does all that without needing to do the copy.  Whether this is useful or not, I don't know, but that's what it's for.

Reply:
Thank you so much for clarification, helps a lot! I don't think that we need it in our use case right now but might need in future.